### ISSUE TITLE
Unable to access Promise polyfill if window.Promise exists

### ISSUE THREAD

**Issue by tivac** [MithrilJS/mithril.js#1960]
## Expected Behavior
I should be able to `require`/`import` mithril's `Promise` polyfill and assign it to `window.Promise` over any existing one, so that I can use a known-good implementation even in environments that may only have a partial or buggy promise implementation.

## Current Behavior
`require`/`import` of `mithril/promise/promise.js` hands me back `window.Promise` or `global.Promise` if they exist, instead of mithril's polyfill.

## Possible Solution
A named export (`exports.polyfill`?) that always provides access to the polyfill.

## Context
I write code in weird environments (embedded web browsers inside of game engines) that often come with buggy or incomplete promise support. I want to use mithril's known-good promise polyfill instead, but am unable to do so from user-land and have to resort to gross hacks to make it available.

**Comment by pygy** [comment]
@tivac couldn't you `delete Promise` before you require Mithril? Or do you have other code that relies on the quirkiness of the default `Promise`?

`exports.polyfill` would b cleaner though.

**Comment by leeoniya** [comment]
> couldn't you delete Promise before you require Mithril

dont think you can delete global vars (or any vars?) in strict mode, only props.

**Comment by tivac** [comment]
@pygy `delete window.Promise` counts as one of those "gross hacks" I referenced above. It does *work*, but it looks super-duper weird IMO.

I freely admit that this is a very edge-case thing, but just bringing it up as something I've run into a couple of times. :sweat_smile:

**Comment by pygy** [comment]
@tivac to each one his own... `exports.polyfill` would actually mean adding a field on the global Promise which is kinda gross as well.

A cleaner solution would be to put the polyfill in `promise/polyfill.js` and keep the code that selects what to exports in `promise/promise.js`... but then `m.request` would still use the broken native module...

So it's not that clean :-/

**Comment by tivac** [comment]
Yeah, making the default export the existing `Promise` object doesn't leave much room for avoiding that. I don't see an obvious win here, in the meantime I've been using a different Promise polyfilly entirely which fills me with sadness because of the waste.

**Comment by spacejack** [comment]
I might have a use for the polyfill as well - I might want to extend the Promise class, but since I'm compiling to ES5 I can't (don't believe it's possible without using ES6 in the browser.) I think I can do that with the polyfill version though.
